File: VB/WinAppCustomCellValue/Form1.vb
Imports DevExpress.XtraEditors
Imports DevExpress.XtraPivotGrid
Imports System.Linq
Imports System

Namespace WinAppCustomCellValue
	Partial Public Class Form1
		Inherits XtraForm
	
		Public Sub New()
			InitializeComponent()
			salesPersonTableAdapter1.Fill(nwindDataSet1.SalesPerson)
			Dim fieldPercentOfBeverages As New PivotGridField() With {.FieldName = "Extended Price", .Area = PivotArea.DataArea, .Caption = "% Beverages", .Name = "PercentOfBeverages"}
			fieldPercentOfBeverages.CellFormat.FormatType = DevExpress.Utils.FormatType.Numeric
			fieldPercentOfBeverages.CellFormat.FormatString = "P"
			pivotGridControl1.Fields.Add(fieldPercentOfBeverages)
		End Sub

		Private Sub pivotGridControl1_CustomCellValue(ByVal sender As Object, ByVal e As PivotCellValueEventArgs) Handles pivotGridControl1.CustomCellValue
			' Calculates the 'Percent' field values.
			If e.DataField.Name = "PercentOfBeverages" Then
				' Do not display grand total values.
				If e.RowValueType = PivotGridValueType.GrandTotal Then
					e.Value = Nothing
					Return
				End If
				Dim rowValues = e.GetRowFields().Select(Function(f) If(f Is fieldCategoryName, "Beverages", e.GetFieldValue(f))).ToArray()
				Dim columnValues = e.GetColumnFields().Select(Function(f) If(f Is fieldCategoryName, "Beverages", e.GetFieldValue(f))).ToArray()
				Dim beveragesValue As Decimal = Convert.ToDecimal(e.GetCellValue(columnValues, rowValues, e.DataField))
				If beveragesValue = 0 Then
					e.Value = Nothing
				Else
					e.Value = Convert.ToDecimal(e.Value) / beveragesValue
				End If
			Else
				Return
			End If
		End Sub
	End Class
End Namespace

File: CS/WinAppCustomCellValue/Form1.cs
using DevExpress.XtraEditors;
using DevExpress.XtraPivotGrid;
using System.Linq;
using System;

namespace WinAppCustomCellValue
{
    public partial class Form1 : XtraForm
    {
        public Form1()
        {
            InitializeComponent();
            salesPersonTableAdapter1.Fill(nwindDataSet1.SalesPerson);

            PivotGridField fieldPercentOfBeverages = new PivotGridField()
            {
                FieldName = "Extended Price",
                Area = PivotArea.DataArea,
                Caption = "% Beverages",
                Name = "PercentOfBeverages"
            };

            fieldPercentOfBeverages.CellFormat.FormatType = DevExpress.Utils.FormatType.Numeric;
            fieldPercentOfBeverages.CellFormat.FormatString = "P";
            pivotGridControl1.Fields.Add(fieldPercentOfBeverages);
        }

        private void pivotGridControl1_CustomCellValue(object sender, PivotCellValueEventArgs e)
        {
            // Calculates the 'Percent' field values.
            if (e.DataField.Name == "PercentOfBeverages")
            {
                // Do not display grand total values.
                if (e.RowValueType == PivotGridValueType.GrandTotal)
                {
                    e.Value = null;
                    return;
                }

                var rowValues = e.GetRowFields().Select(f => f.FieldName == "CategoryName" ? "Beverages" : e.GetFieldValue(f)).ToArray();
                var columnValues = e.GetColumnFields().Select(f => f.FieldName == "CategoryName" ? "Beverages" : e.GetFieldValue(f)).ToArray();
                decimal beveragesValue = Convert.ToDecimal(e.GetCellValue(columnValues, rowValues, e.DataField));
                if (beveragesValue == 0)
                    e.Value = null;
                else
                    e.Value = Convert.ToDecimal(e.Value) / beveragesValue;
            }
            else return;
        }
    }
}
